### PR TITLE
Adding VM support for wait/resume and async invocations.

### LINKED ISSUE #9503: Support manual management of VM stack tracing zones.
Coroutines in #8093 require the `IREE_TRACING_FEATURE_FIBERS` flag to be set today as when we unwind the host stack the zones remain open. Adding a top-level enter/leave that calls the IREE_TRACE_FIBER_ENTER/LEAVE when enabled and otherwise ends and begins the zones for the whole stack would let us preserve proper nesting and run with the fiber feature flag off.

I haven't done any profiling to see how bad the flag is and just noted in the tracy docs that it "adds overhead" - looks like it may change from a lock-free queue to a locking one but that may not matter much for us. So long as we have the feature flag option though it'd be good to work with it disabled.

### PR DESCRIPTION
**Pull request by benvanik**
The two major additions are the low-level `iree_vm_begin/resume/end_invoke` methods and the high-level `iree_vm_async_invoke` method. The existing `iree_vm_invoke` has been rewritten to reuse the `iree_vm_*_invoke` methods such that all invocations are handled in a consistent way. Bindings can choose to use either set of methods. The low-level methods don't dictate a scheduling mechanism and can tightly integrate with external schedulers. The high-level method uses the `iree_loop_t` interface which can be mapped to normal event loop mechanisms and will handle state tracking. In either case the binding layer provides the storage and lifetime management and both sets of methods don't allocate any additional memory.

This PR only contains the VM portions of the async scheduling work: the stream->HAL compiler lowering and runtime module need to be adapted to use semaphores and there are some TODOs around semaphore exports to wait handles that need to be done. Experiments show that about ~100 LOC is all that's needed to adapt `iree-run-module` to perform concurrent execution via `iree_vm_async_invoke` and (when the TODOs are worked around) the HAL can yield:
image

Most of the implementation work here was done years ago by having the VM module interface expose the begin_call and resume_call. This set of changes completes that work for bytecode and native modules, but emitc is going to need additional work to handle yielding. Future changes that enable the yielding behavior in the compiler and HAL module will have flags to disable them for when emitc is used. Documented in #9612.

Progress on #8093 (phase 1 complete).
Fixes #9503.

### CODE REVIEW

File: runtime/src/iree/vm/stack.c [hunk 1]
```diff
@@ -551,3 +692,74 @@ IREE_API_EXPORT iree_status_t iree_vm_stack_annotate_backtrace(
...
+
+#if (IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION) && \
+    !(IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_FIBERS)
+
+// These helpers perform manual fiber-like stack suspend/resume when not
+// natively supported by the tracing library. See
+// iree_vm_stack_suspend_trace_zones for more information.
+
+IREE_API_EXPORT void iree_vm_stack_suspend_trace_zones(iree_vm_stack_t* stack) {
+  // Walk top->bottom to unwind the trace zones.
```
Comment (stellaraccident) [inline runtime/src/iree/vm/stack.c:704]: Ugh. (no better idea)

**Review by stellaraccident — APPROVED** [review]
This looks approximately like I expected. Glad you wrote it and not me: I've done this before and it is some hard programming. A handful of comments, but I didn't see any flaws. Just generally would recommend some asserts/something about invariants: these things can be really hard to debug when they go wrong.

File: runtime/src/iree/vm/invocation.c [hunk 2]
```diff
@@ -149,89 +180,766 @@ static iree_status_t iree_vm_invoke_marshal_outputs(
...
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+//===----------------------------------------------------------------------===//
+// Asynchronous invocation
+//===----------------------------------------------------------------------===//
+
+// WARNING: this function cannot have any trace markers that span the begin
```
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/src/iree/vm/invocation.c:354]: Unfortuantely not though perhaps we could build something. Tracy's server (recording) side does not track zone nesting so the only thing that can reveal incorrect behavior is the client (profiler) that does the analysis. It'll pop up an error indicating which zone was incorrectly nested but it does require that a user is actually tracing to see it.

We could have an expensive asserts check that would keep a local zone stack but we'd need to support tracking fiber lifetimes and such (as each thread and fiber have their own stacks) - I'm not quite sure how to do that yet but maybe it's something that could be done easier in the tracy server code. Probably worth an upstream suggestion.

File: runtime/src/iree/vm/invocation.c [hunk 3]
```diff
@@ -149,89 +180,766 @@ static iree_status_t iree_vm_invoke_marshal_outputs(
...
+//        as the trace stack is suspended/resumed and zones get extra counts.
+//
+// To make concurrent coroutines work when Tracy's fiber support is enabled we
+// go from treating each context as a fiber to treating each invocation as one.
+// This has the side-effect of creating one fiber per invocation and in
+// benchmarks that can be really noisy; best that can be done there is disabling
+// native fiber support.
+static iree_vm_invocation_id_t iree_vm_invoke_allocate_id(
+    iree_vm_context_t* context, const iree_vm_function_t* function) {
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_FIBERS
```
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/src/iree/vm/invocation.c:201]: Yep - tracy requires persistent addresses that outlive the trace server. Simplifies a lot but does lead to things like this. I take solace in at least having been able to sequester this to a single location :)

File: runtime/src/iree/vm/stack.c [hunk 4]
```diff
@@ -290,9 +283,21 @@ iree_vm_stack_invocation_flags(const iree_vm_stack_t* stack) {
   return stack->flags;
 }
 
-IREE_API_EXPORT iree_vm_context_id_t
-iree_vm_stack_context_id(const iree_vm_stack_t* stack) {
-  return stack->context_id;
+IREE_API_EXPORT iree_vm_stack_frame_t* iree_vm_stack_top(
```
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/src/iree/vm/stack.c:286]: Ensured there was failure testing at the call-sites and added asserts where they shouldn't be needed.

**Pull request merged**